lib/grandpa/message_handler.go

```diff
@@ -87,18 +96,58 @@ func (h *MessageHandler) handleNeighbourMessage(msg *NeighbourMessage) error {
 
 	// TODO; determine if there is some reason we don't receive justifications in responses near the head (usually),
 	// and remove the following code if it's fixed. (#1815)
-	head, err := h.blockState.BestBlockNumber()
+	// head, err := h.blockState.BestBlockNumber()
+	// if err != nil {
+	// 	return err
+	// }
+
+	// TODO: Why are we ignoring these? Isn't
+	// msg.Number likely to be higher if we are lagging behind?
+	// ignore neighbour messages that are above our head
+	// if int64(msg.Number) > head.Int64() {
+	// 	return nil
+	// }
+
+	logger.Debugf("got neighbour message with number %d, set id %d and round %d, from: %s ",
+		msg.Number, msg.SetID, msg.Round, from)
+	// TODO: should we send a justification request here? potentially re-connect this to sync package? (#1815)
+
+	highestRound, setID, err := h.blockState.GetHighestRoundAndSetID()
 	if err != nil {
 		return err
 	}
 
-	// ignore neighbour messages that are above our head
-	if int64(msg.Number) > head.Int64() {
-		return nil
+	// catch up only if we are behind by more than catchup threshold
+	if (int(msg.Round) - int(highestRound)) > catchupThreshold {
```

Comment:
if our set IDs are different, this may not work correctly. round resets to 0 when the set ID changes, so I'd recommend checking that the set IDs match first before doing this

Reply:
added set id check